docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerSlave.java

```diff
@@ -135,43 +147,101 @@ private void slaveShutdown(TaskListener listener) throws DockerException, IOExce
+                    ));
+        
+        // Tag latest if specified
+        if(getJobProperty().isTagLatest()) {
+            imageTags.add("latest");
+        }
+
+        // If there's no tags, or if specified, tag with the build number
+        if(imageTags.isEmpty() || getJobProperty().isTagBuildNumber()) {
+            imageTags.add(theRun.getDisplayName());
+        }
 
-         // Commit
-        String tag_image = client.commitCmd(containerId)
-                    .withRepository(theRun.getParent().getDisplayName())
-                    .withTag(theRun.getDisplayName())
-                    .withAuthor("Jenkins")
+        // Commit image and get new image ID
+        String imageId = client.commitCmd(containerId)
+                    .withAuthor(getJobProperty().getImageAuthor())
                     .exec();
+        LOGGER.log(Level.INFO, "Commited to image: " + imageId);
 
-        // Tag it with the jenkins name
-        addJenkinsAction(tag_image);
+        // Iterate list of tags, and tag appropriately
+        for(String tag : imageTags) {
+            // Clean up the image tag
+            String cleanTag = cleanImageTag(tag);
 
-        // SHould we add additional tags?
-        try
-        {
-            String tagToken = getAdditionalTag(listener);
+            // Skip empty tags
+            if(Strings.isNullOrEmpty(cleanTag)){
+                continue;
+            }
+            LOGGER.log(Level.INFO, "Tagging with: " + cleanTag);
 
-            if( !Strings.isNullOrEmpty(tagToken) ) {
-                // ?? client.image(tag_image).tag(tagToken, false);
-                client.tagImageCmd(tag_image,null,tagToken).exec();
-                addJenkinsAction(tagToken);
+            // Tag the image!
+            try {
+                client.tagImageCmd(imageId, repositoryName, cleanTag)
+                    .withForce()
+                    .exec();
+                // addJenkinsAction(cleanTag);
+            } catch(Exception ex) {
+                LOGGER.log(Level.SEVERE, "Could not add tag: " + cleanTag, ex);
+            }
 
-                if( getJobProperty().pushOnSuccess ) {
-                    client.pushImageCmd(tagToken).exec();
+            // Push the image to the registry with all of the tags
+            if(getJobProperty().isPushOnSuccess()) {
+                try {
+                    client.pushImageCmd(repositoryName)
+                        .withTag(cleanTag)
+                        .exec();
+                } catch (NotFoundException ex) {
+                    // This should NEVER happen but just in case.
+                    LOGGER.log(Level.SEVERE, "Couldnt find " + repositoryName + ":" +cleanTag + ". Not pushing.");
                 }
             }
         }
-        catch(Exception ex) {
-            LOGGER.log(Level.SEVERE, "Could not add additional tags");
-        }
 
-        if( getJobProperty().cleanImages ) {
+        // Add a Jenkins 'Action' for the newly committed image
```

Comment:
Yeah, it takes the ID from the newly committed image and does a `docker rmi -f ...`